### ISSUE TITLE
Add something similar to React's shouldComponentUpdate

### ISSUE THREAD

**Issue by StokeMasterJack** [flutter/flutter#19421]
It would be nice if there was a way to tell a component to not re-render if data has not changed.

**Comment by zoechi** [comment]
Can you please provide more information about a concrete use case?

**Comment by StokeMasterJack** [comment]
It's the same use-case as for React. 

Suppose you have a deep tree of widgets. And suppose there is a state change in one one the widgets near the top of the tree. This would, of course, trigger that widget to re-render. And, indirectly, it would also trigger a re-render for all descendent widgets.

But suppose that one of those descendants doesn't need to re-render. Because his state didn't change.  In React, there is callback on the widget, called `shouldComponentUpdate` that allows a child, essentially, to say "i don't need to re-render". 

`shouldComponentUpdate` is passed in the oldState and the newState. And that is what it uses to make the determination:

```
bool shouldComponentUpdate(oldState, newtState){
  return oldState != newtState;
}
```

**Comment by zoechi** [comment]
@StokeMasterJack I have heard of react already ;-)

Flutter should take care of that automatically.
Perhaps @Hixie can explain that with more confidence.

**Comment by Hixie** [comment]
It's a pretty good idea, in fact you'll see that we use it extensively among the delegate classes.

For widgets it's a little tricky because to implement correctly you'd also have to call it for your "child" widget (since while you may not need updating, you may need to rebuild to provide the new "child" widget if it _does_ need updating), but then you'd end up walking the whole tree at each level to determine if any level should be skipped, which ends up being O(N^2).

What we do instead is if you want to make sure that your subtree isn't updated, just return the same instance as you did the previous time. If we find that `build` returned the same instance both times, we short-circuit the walk and do no work for that subtree. (And there's no need to walk the subtree because if a widget is the same instance as its previous version, then all the children will be too, since widgets are immutable.) So this ends up being much more efficient.

**Comment by zoechi** [comment]
@Hixie interesting. I had expected returning the same instance to cause troubles because a stale context was used to build it.

**Comment by Hixie** [comment]
It's common to return a `const` widget, which is literally the same instance across the entire lifetime of the app.

If the instance depends on the context in some way (e.g. the parent used `MediaQuery.of` or `Theme.of` in its construction) then it should not be cached so long as those values haven't changed, that is true. But then "shouldComponentUpdate" would have to return true in those cases too; the result is the same.

**Comment by zoechi** [comment]
Is there still something left to do or can the issue be closed?

**Comment by zoechi** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. We are therefore reluctantly going to close this bug for now. Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!
Thanks for your contribution.

**Comment by bendelonlee** [comment]
I was wondering about this, ran across this old issue while Googling. Then, I asked this question on SO and got some answers: https://stackoverflow.com/questions/67358514/how-to-prevent-unnecessary-renders-in-flutter/67376874#67376874

TLDR: Flutter is very optimized out-of-the-box but if you're looking to remove useless rebuilds you can use BLOC's `buildWhen`, which is very similar to `shouldComponentUpdate`

Cheers!
